extra_data/sourcedata.py

```diff
@@ -0,0 +1,81 @@
+from typing import List
+
+import h5py
+
+from .exceptions import PropertyNameError
+from .file_access import FileAccess
+from .keydata import KeyData
+
+class SourceData:
+    """Data for one key in one source
+
+    Don't create this directly; get it from ``run[source]``.
+    """
+    def __init__(
+            self, source, *, sel_keys, train_ids, files, section,
+            inc_suspect_trains=True,
+    ):
+        self.source = source
+        self.sel_keys = sel_keys
+        self.train_ids = train_ids
+        self.files: List[FileAccess] = files
+        self.section = section
+        self.inc_suspect_trains = inc_suspect_trains
+
+    def __repr__(self):
+        return f"<extra_data.SourceData source={self.source!r} " \
+               f"for {len(self.train_ids)} trains>"
+
+    def _has_exact_key(self, key):
+        if self.sel_keys is not None:
+            return key in self.sel_keys
+
+        for f in self.files:
+            return f.has_source_key(self.source, key)
+
+    def __contains__(self, key):
+        res = self._has_exact_key(key)
+        if (not res) and (self.section == 'CONTROL'):
+            res = self._has_exact_key(key + '.value')
+        return res
+
+    def __getitem__(self, key):
+        if key not in self:
+            raise PropertyNameError(key, self.source)
+        ds0 = self.files[0].file[
+            f"{self.section}/{self.source}/{key.replace('.', '/')}"
+        ]
+        if isinstance(ds0, h5py.Group):
+            # This can only occur with a CONTROL key missing its .value suffix
+            ds0 = ds0['value']
+            key += '.value'
+
+        return KeyData(
+            self.source,
+            key,
+            train_ids=self.train_ids,
+            files=self.files,
+            section=self.section,
+            dtype=ds0.dtype,
+            eshape=ds0.shape[1:],
+            inc_suspect_trains=self.inc_suspect_trains,
+        )
+
+    def keys(self):
```

Comment:
I'm sorry striking the same notch (this saying probably doesn't work in English), but what do you think about adding an optional flag here to collapse `.value` and `.timestamp`? Something like `omit_suffix` or `collapse_value_and_timestamp`.

While it might only be useful in generic programming, I could imagine it something people want to do with the output from this function, and it might be convenient to provide it.

Reply:
That seems reasonable enough. What about something that defaults to True for including both as separate keys, like `inc_timestamps=False` or `value_and_timestamp=False`?

Reply:
Sure, either way is fine with me!